src/pwm.rs

```diff
@@ -12,98 +12,136 @@ use core::time::Duration;
+///
 /// }
 ///
 /// # use core::convert::Infallible;
 /// # trait U32Ext { fn khz(self) -> Duration; }
 /// # impl U32Ext for u32 { fn khz(self) -> Duration { Duration::from_nanos( 1_000_000u64 / self as u64 ) } }
-/// # enum Channel { _1, _2 }
-/// # struct Pwm1;
-/// # impl hal::pwm::Pwm for Pwm1 {
+/// // Note: possibly you could genericise this over [PwmPinImpl; CHANNELS], which would be great for accessing channels internally
+/// // except, these are probably not all going to be the same type, so instead we're using marker traits to specify whether it has been split or not
+/// # struct Pwm1<S>(S);
+/// # struct Pwm1Pin;
+/// # struct HasPins;
+/// # impl Pwm1<HasPins> {
+/// #     pub fn split(self) -> (Pwm1<()>, [Pwm1Pin; 2]) {
+/// #       (Pwm1(()), [Pwm1Pin, Pwm1Pin])
+/// #     }
+/// # }
+/// # impl Pwm1<()> {
+/// #     pub fn join(self, pins: [Pwm1Pin; 2]) -> Pwm1<HasPins> {
+/// #       Pwm1(HasPins)
+/// #     }
+/// # }
+/// #
+/// # impl <S> hal::pwm::Pwm<u16> for Pwm1<S> {
 /// #     const CHANNELS: usize = 4;
+/// #     const MAX_DUTY: u16 = 1024;
 /// #     type Error = Infallible;
-/// #     type Duty = u16;
+/// #
 /// #     fn try_disable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
 /// #     fn try_enable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
-/// #     fn try_get_max_duty(&self) -> Result<u16, Self::Error> { Ok(0) }
-/// #     fn try_set_duty(&mut self, _: usize, _: u16) -> Result<(), Self::Error> { Ok(()) }
 /// #     fn try_get_period(&self) -> Result<Duration, Self::Error> { unimplemented!() }
 /// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<Duration> { Ok(()) }
 /// # }
+/// #
+/// # impl hal::pwm::PwmDuty<u16> for Pwm1<HasPins> {
+/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: usize, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
+/// # impl hal::pwm::PwmPin<u16> for Pwm1Pin {
+/// #     const MAX_DUTY: u16 = 1024;
+/// #     type Error = Infallible;
+/// #     fn try_disable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_enable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_get_duty(&self) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
 /// ```
 // unproven reason: pre-singletons API. The `PwmPin` trait seems more useful because it models independent
 // PWM channels. Here a certain number of channels are multiplexed in a single implementer.
-pub trait Pwm {
+pub trait Pwm<Duty> {
     /// Number of available channels
     const CHANNELS: usize;
 
+    /// Maximum duty cycle value
+    const MAX_DUTY: Duty;
```

Comment:
I'm wondering how much good that will do. Any driver built on top of this still won't know what `Duty` actually is, so what can it do with this constant? Maybe we should have a duty `Duty` trait that is implemented for any type that provides certain operations (division, multiplication)? Then it would be pretty straight-forward to do something like `set_duty(MAX_DUTY / 2)`.

Reply:
a driver built on this _has_ to know what duty is (or have a bound supporting conversion to duty) or it can't call any of the methods?

it would be possible for drivers to impose `num_traits` bounds on `Duty` if required, or to directly implement over `<u16>` and `<u32>` if those are all the duty types we see in practice. is there a case in which it's useful to a HAL to implement multiple `Duty` instances or could this be simplified by assuming Duty is a timer-defined primitive?

Reply:
> a driver built on this _has_ to know what duty is (or have a bound supporting conversion to duty) or it can't call any of the methods?

I can imagine cases where all duty values are passed into the driver by the consumer, but even if that is a real-word use case, it would be limited.

> it would be possible for drivers to impose `num_traits` bounds on `Duty` if required,

I agree that this is totally reasonable. This is another point where I'm concerned about stabilization though. If it turned out that most drivers require similar bounds, and HALs can provide them (in the end, duty is just a number after all, or isn't it?), we could later add those bounds to embedded-hal, to improve interoperability.

With 1.0 out, we could lock ourselves into a non-optimal situation, where most drivers require some bounds, but not always the same ones, and not all HALs provide them (if the duty type were a newtype, and not `u16`/`u32` directly, for example).

> or to directly implement over `<u16>` and `<u32>` if those are all the duty types we see in practice.

I think this would be an unfortunate situation. It's fine for SPI/Serial, as there's a standard type that every HAL provides in practice (`u8`), but I don't see such a standard type here (as timers have different resolutions). Not sure what the best solution here is.

> is there a case in which it's useful to a HAL to implement multiple `Duty` instances or could this be simplified by assuming Duty is a timer-defined primitive?

I don't know. If stabilization weren't a concern (and it wouldn't be, if we had embedded-pwm), I'd suggest to assume that `Duty` is timer-defined, make it an associated type, and wait for the complaints to roll in.